File: src/test/java/no/novari/flyt/integration/validation/constraints/ActiveConfigurationIsDefinedIfStateIsActiveValidatorTest.java
package no.novari.flyt.integration.validation.constraints;

import no.novari.flyt.integration.model.dtos.ConfigurationDto;
import no.novari.flyt.integration.model.entities.Integration;
import no.novari.flyt.integration.validation.IntegrationValidationContext;
import org.hibernate.validator.constraintvalidation.HibernateConstraintValidatorContext;
import org.hibernate.validator.internal.engine.constraintvalidation.ConstraintValidatorContextImpl;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class ActiveConfigurationIsDefinedIfStateIsActiveValidatorTest {

    private ActiveConfigurationIsDefinedIfStateIsActiveValidator validator;
    private HibernateConstraintValidatorContext context;
    private IntegrationValidationContext validationContext;

    @BeforeEach
    public void setUp() {
        validator = new ActiveConfigurationIsDefinedIfStateIsActiveValidator();
        context = mock(ConstraintValidatorContextImpl.class);
        validationContext = mock(IntegrationValidationContext.class);
    }

    @Test
    public void shouldReturnTrueWhenStateIsNotActive() {
        when(context.getConstraintValidatorPayload(IntegrationValidationContext.class))
                .thenReturn(validationContext);

        boolean result = validator.isValid(Integration.State.DEACTIVATED, context);

        assertTrue(result);
    }

    @Test
    public void shouldReturnTrueWhenStateIsActiveAndConfigurationIsNotNull() {
        when(context.getConstraintValidatorPayload(IntegrationValidationContext.class))
                .thenReturn(validationContext);
        when(validationContext.getConfiguration()).thenReturn(
                ConfigurationDto.builder()
                        .id(1L)
                        .integrationId(2L)
                        .integrationMetadataId(3L)
                        .completed(true)
                        .comment("Test Comment")
                        .version(1)
                        .build());

        boolean result = validator.isValid(Integration.State.ACTIVE, context);

        assertTrue(result);
    }

    @Test
    public void shouldReturnFalseWhenStateIsActiveAndConfigurationIsNull() {
        when(context.getConstraintValidatorPayload(IntegrationValidationContext.class))
                .thenReturn(validationContext);
        when(validationContext.getConfiguration()).thenReturn(null);

        boolean result = validator.isValid(Integration.State.ACTIVE, context);

        assertFalse(result);
    }

}

File: src/test/kotlin/no/novari/flyt/integration/validation/constraints/ActiveConfigurationIsDefinedIfStateIsActiveValidatorTest.kt
package no.novari.flyt.integration.validation.constraints

import no.novari.flyt.integration.api.dto.ConfigurationDto
import no.novari.flyt.integration.persistence.entity.Integration
import no.novari.flyt.integration.validation.IntegrationValidationContext
import org.hibernate.validator.constraintvalidation.HibernateConstraintValidatorContext
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.kotlin.mock
import org.mockito.kotlin.whenever

class ActiveConfigurationIsDefinedIfStateIsActiveValidatorTest {
    private lateinit var validator: ActiveConfigurationIsDefinedIfStateIsActiveValidator
    private lateinit var context: HibernateConstraintValidatorContext
    private lateinit var validationContext: IntegrationValidationContext

    @BeforeEach
    fun setUp() {
        validator = ActiveConfigurationIsDefinedIfStateIsActiveValidator()
        context = mock()
        validationContext = mock()
    }

    @Test
    fun `returns true when state is not active`() {
        whenever(context.getConstraintValidatorPayload(IntegrationValidationContext::class.java))
            .thenReturn(validationContext)

        val result = validator.isValid(Integration.State.DEACTIVATED, context)

        assertTrue(result)
    }

    @Test
    fun `returns true when state is active and configuration is present`() {
        whenever(context.getConstraintValidatorPayload(IntegrationValidationContext::class.java))
            .thenReturn(validationContext)
        whenever(validationContext.configuration)
            .thenReturn(
                ConfigurationDto(
                    id = 1L,
                    integrationId = 2L,
                    integrationMetadataId = 3L,
                    completed = true,
                    comment = "Test Comment",
                    version = 1,
                ),
            )

        val result = validator.isValid(Integration.State.ACTIVE, context)

        assertTrue(result)
    }

    @Test
    fun `returns false when state is active and configuration is missing`() {
        whenever(context.getConstraintValidatorPayload(IntegrationValidationContext::class.java))
            .thenReturn(validationContext)
        whenever(validationContext.configuration).thenReturn(null)

        val result = validator.isValid(Integration.State.ACTIVE, context)

        assertFalse(result)
    }
}
